File: apps/desktop/src/stores/universalPlayerStore.ts
import { create } from 'zustand';
import { useVlcPlayerStore } from './vlcPlayer';
import { useP2PPlayerStore } from './p2pPlayerStore';
import { useP2PStore } from './p2pStore';
import { WatchableObject } from '@zenith-tv/content';
import type {
  VlcState,
  VlcTrack,
  ScreenMode,
  OpenOptions,
  PlaybackOptions,
  AudioOptions,
  VideoOptions,
  SubtitleOptions,
  WindowOptions,
  ShortcutOptions,
} from '@zenith-tv/content';

// Define the interface (same as VlcPlayerState)
interface UniversalPlayerState {
  // State
  isAvailable: boolean;
  isInitialized: boolean;
  playerState: VlcState;
  time: number;
  duration: number;
  volume: number;
  isMuted: boolean;
  noAudio: boolean | undefined;
  audioTracks: VlcTrack[];
  subtitleTracks: VlcTrack[];
  videoTracks: VlcTrack[];
  currentAudioTrack: number;
  currentSubtitleTrack: number;
  currentVideoTrack: number;
  error: string | null;

  // Playback info
  position: number;
  buffering: number;
  rate: number;
  isSeekable: boolean;

  // Video settings
  aspectRatio: string | null;
  crop: string | null;
  scale: number;
  deinterlace: string | null;

  // Delay settings
  audioDelay: number;
  subtitleDelay: number;

  // Screen mode
  screenMode: ScreenMode;
  prevScreenMode: ScreenMode;
  stickyElement: HTMLElement | null;
  wasPlayingBeforeMinimize: boolean;
  currentItem: WatchableObject | null;

  // Actions
  init: () => Promise<void>;
  play: (item: WatchableObject) => Promise<void>;
  setScreenMode: (mode: ScreenMode) => void;
  setStickyElement: (element: HTMLElement | null) => void;

  // Unified API
  open: (options: OpenOptions | string) => Promise<void>;
  playback: (options: PlaybackOptions) => Promise<void>;
  audio: (options: AudioOptions) => Promise<void>;
  video: (options: VideoOptions) => Promise<void>;
  subtitle: (options: SubtitleOptions) => Promise<void>;
  window: (options: WindowOptions) => Promise<boolean>;
  shortcut: (options: ShortcutOptions) => Promise<void>;

  // Helpers
  shouldStickyPanelVisible: () => boolean;
}

// Active-store selection is driven by P2P mode, not by selectedDeviceId.
//
//  - mode === 'server'  → p2pPlayerStore  (this desktop is a remote control;
//                                          no local VLC, mirrors a client)
//  - otherwise          → vlcPlayerStore  ('off' or 'client' — VLC runs here)
//
// In client mode the VLC store handles both the local UI and inbound P2P
// commands, so there's only ever one real source of truth at a time.
const isServerMode = () => useP2PStore.getState().mode === 'server';

const getActiveStore = () =>
  isServerMode() ? useP2PPlayerStore.getState() : useVlcPlayerStore.getState();

// Project just the *state* fields out of a source store snapshot, so we
// don't accidentally overwrite the universal store's action functions
// when we `set()` from a subscribe callback. Zustand's set is a shallow
// merge, and without this filter the source store's actions (which close
// over the wrong store!) would replace our proxy actions.
type PlayerStateFields = Omit<
  UniversalPlayerState,
  | 'init' | 'play' | 'setScreenMode' | 'setStickyElement'
  | 'open' | 'playback' | 'audio' | 'video' | 'subtitle' | 'window' | 'shortcut'
  | 'shouldStickyPanelVisible'
>;
type AnyPlayerSnapshot = Record<string, unknown>;
const pickState = (src: AnyPlayerSnapshot): PlayerStateFields => ({
  isAvailable: src.isAvailable as boolean,
  isInitialized: src.isInitialized as boolean,
  playerState: src.playerState as VlcState,
  time: src.time as number,
  duration: src.duration as number,
  volume: src.volume as number,
  isMuted: src.isMuted as boolean,
  noAudio: src.noAudio as boolean | undefined,
  audioTracks: src.audioTracks as VlcTrack[],
  subtitleTracks: src.subtitleTracks as VlcTrack[],
  videoTracks: src.videoTracks as VlcTrack[],
  currentAudioTrack: src.currentAudioTrack as number,
  currentSubtitleTrack: src.currentSubtitleTrack as number,
  currentVideoTrack: src.currentVideoTrack as number,
  error: src.error as string | null,
  position: src.position as number,
  buffering: src.buffering as number,
  rate: src.rate as number,
  isSeekable: src.isSeekable as boolean,
  aspectRatio: src.aspectRatio as string | null,
  crop: src.crop as string | null,
  scale: src.scale as number,
  deinterlace: src.deinterlace as string | null,
  audioDelay: src.audioDelay as number,
  subtitleDelay: src.subtitleDelay as number,
  screenMode: src.screenMode as ScreenMode,
  prevScreenMode: src.prevScreenMode as ScreenMode,
  stickyElement: src.stickyElement as HTMLElement | null,
  wasPlayingBeforeMinimize: src.wasPlayingBeforeMinimize as boolean,
  currentItem: src.currentItem as WatchableObject | null,
});

export const useUniversalPlayerStore = create<UniversalPlayerState>((set) => {
  // Ensure the P2P mirror listeners are live from the start so we don't
  // miss state_update packets that arrive before the user opens the UI.
  useP2PPlayerStore.getState()._setupListeners();

  // Re-sync whenever P2P mode flips (off ↔ client ↔ server).
  useP2PStore.subscribe((state, prevState) => {
    if (state.mode !== prevState.mode) {
      set(pickState(getActiveStore() as unknown as AnyPlayerSnapshot));
    }
  });

  // Mirror local VLC store into the universal facade when we're NOT in
  // server mode. (Client mode also uses vlcPlayerStore locally.)
  useVlcPlayerStore.subscribe((state) => {
    if (!isServerMode()) set(pickState(state as unknown as AnyPlayerSnapshot));
  });

  // Mirror P2P player store into the universal facade when we ARE in
  // server mode — it's being fed by vlc_event packets from the client.
  useP2PPlayerStore.subscribe((state) => {
    if (isServerMode()) set(pickState(state as unknown as AnyPlayerSnapshot));
  });

  return {
    // Start from the VLC store as the default snapshot.
    ...pickState(useVlcPlayerStore.getState() as unknown as AnyPlayerSnapshot),

    // Actions proxy to whichever store is active at call time.
    init: async () => getActiveStore().init(),
    play: async (item) => getActiveStore().play(item),
    setScreenMode: (mode) => getActiveStore().setScreenMode(mode),
    setStickyElement: (element) => getActiveStore().setStickyElement(element),
    open: async (opts) => getActiveStore().open(opts),
    playback: async (opts) => getActiveStore().playback(opts),
    audio: async (opts) => getActiveStore().audio(opts),
    video: async (opts) => getActiveStore().video(opts),
    subtitle: async (opts) => getActiveStore().subtitle(opts),
    window: async (opts) => getActiveStore().window(opts),
    shortcut: async (opts) => getActiveStore().shortcut(opts),
    shouldStickyPanelVisible: () => getActiveStore().shouldStickyPanelVisible(),
  };
});

File: apps/mobile/lib/stores/universal_player_store.dart
import 'package:flutter/foundation.dart';
import '../core/app_logger.dart';
import '../models/watchable.dart';
import '../p2p/models/client_event.dart';
import '../p2p/models/p2p_message.dart';
import 'media_player_store.dart';
import 'remote_player_store.dart';

/// P2P mode values — mirrors apps/desktop/src/stores/p2pStore.ts
enum P2PMode { off, client, server }

/// Facade store that delegates to the active player source based on P2P mode:
///
///   off / client → MediaPlayerStore (local media_kit, source of truth)
///   server       → RemotePlayerStore (mirror fed by client_event packets)
///
/// Commands in server mode are forwarded over P2P via [sendP2PCommand].
/// Components read from this store regardless of which backend is active.
/// Mirrors: apps/desktop/src/stores/universalPlayerStore.ts
class UniversalPlayerStore extends ChangeNotifier {
  final MediaPlayerStore localPlayer;
  final RemotePlayerStore remotePlayer;

  /// Injected by the app shell — routes commands to the connected P2P client.
  /// Called only when mode == server.
  void Function(String type, Map<String, dynamic> payload)? sendP2PCommand;

  P2PMode _mode = P2PMode.off;
  P2PMode get mode => _mode;

  bool get _isServerMode => _mode == P2PMode.server;

  UniversalPlayerStore({
    required this.localPlayer,
    required this.remotePlayer,
  }) {
    localPlayer.addListener(_onSourceChanged);
    remotePlayer.addListener(_onSourceChanged);
  }

  void _onSourceChanged() => notifyListeners();

  // ---------------------------------------------------------------------------
  // Mode switching
  // ---------------------------------------------------------------------------

  void setMode(P2PMode mode) {
    if (_mode == mode) return;
    _mode = mode;
    notifyListeners();
  }

  // ---------------------------------------------------------------------------
  // State — proxied from the active source
  // ---------------------------------------------------------------------------

  String get playerState =>
      _isServerMode ? remotePlayer.playerState : localPlayer.playerState;

  double get time =>
      _isServerMode ? remotePlayer.time : localPlayer.time;

  double get duration =>
      _isServerMode ? remotePlayer.duration : localPlayer.duration;

  double get position =>
      _isServerMode ? remotePlayer.position : localPlayer.position;

  double get buffering =>
      _isServerMode ? remotePlayer.buffering : localPlayer.buffering;

  double get volume =>
      _isServerMode ? remotePlayer.volume : localPlayer.volume;

  bool get isMuted =>
      _isServerMode ? remotePlayer.isMuted : localPlayer.isMuted;

  double get rate =>
      _isServerMode ? remotePlayer.rate : localPlayer.rate;

  bool get isSeekable =>
      _isServerMode ? remotePlayer.isSeekable : localPlayer.isSeekable;

  String? get error =>
      _isServerMode ? remotePlayer.error : localPlayer.error;

  String? get currentUrl =>
      _isServerMode ? remotePlayer.currentUrl : localPlayer.currentUrl;

  // currentItem and isFullscreen are local-only concerns
  WatchableObject? get currentItem => localPlayer.currentItem;
  bool get isFullscreen => localPlayer.isFullscreen;

  List<VlcTrack> get audioTracks =>
      _isServerMode ? remotePlayer.audioTracks : localPlayer.audioTracks;

  List<VlcTrack> get subtitleTracks =>
      _isServerMode ? remotePlayer.subtitleTracks : localPlayer.subtitleTracks;

  int get currentAudioTrack =>
      _isServerMode ? remotePlayer.currentAudioTrack : localPlayer.currentAudioTrack;

  int get currentSubtitleTrack =>
      _isServerMode ? remotePlayer.currentSubtitleTrack : localPlayer.currentSubtitleTrack;

  // ---------------------------------------------------------------------------
  // Commands
  //
  //   local / client mode → MediaPlayerStore (drives media_kit directly)
  //   server mode         → sendP2PCommand (relays to connected player device)
  // ---------------------------------------------------------------------------

  /// Play a watchable item. In server mode: sets currentItem locally (for UI)
  /// and sends open command to the connected client over P2P.
  Future<void> play(WatchableObject item) async {
    AppLogger.app('play: ${item.name} [${item.url}] mode=$_mode');
    await localPlayer.open(item.url, item: item);
    if (_isServerMode) {
      _send(P2PMessageType.open, {'file': item.url});
    }
  }

  /// Stop playback and clear currentItem — closes the player panel.
  Future<void> close() async {
    AppLogger.app('close: mode=$_mode');
    await localPlayer.close();
    if (_isServerMode) {
      _send(P2PMessageType.playback, {'action': 'stop'});
    }
  }

  Future<void> open(String url, {WatchableObject? item}) async {
    if (_isServerMode) {
      _send(P2PMessageType.open, {'file': url});
    } else {
      await localPlayer.open(url, item: item);
    }
  }

  Future<void> playback({
    String? action,
    double? time,
    double? position,
    double? rate,
  }) async {
    if (_isServerMode) {
      _send(P2PMessageType.playback, {
        'action': action, 'time': time, 'position': position, 'rate': rate,
      });
    } else {
      await localPlayer.playback(
        action: action,
        time: time,
        position: position,
        rate: rate,
      );
    }
  }

  Future<void> audio({double? volume, bool? mute, int? track}) async {
    if (_isServerMode) {
      _send(P2PMessageType.audio, {'volume': volume, 'mute': mute, 'track': track});
    } else {
      await localPlayer.audio(volume: volume, mute: mute, track: track);
    }
  }

  Future<void> subtitle({int? track}) async {
    if (_isServerMode) {
      _send(P2PMessageType.subtitle, {'track': track});
    } else {
      await localPlayer.subtitle(track: track);
    }
  }

  void setFullscreen(bool value) => localPlayer.setFullscreen(value);

  void _send(String type, Map<String, dynamic> payload) {
    final filtered = {
      for (final e in payload.entries)
        if (e.value != null) e.key: e.value,
    };
    AppLogger.p2p('→ $type $filtered');
    sendP2PCommand?.call(type, filtered);
  }

  // ---------------------------------------------------------------------------
  // P2P helpers
  // ---------------------------------------------------------------------------

  /// Called by P2PManager when a client_event arrives in server mode.
  void applyClientEvent(ClientEventData event) {
    remotePlayer.applyClientEvent(event);
  }

  /// Called by P2PManager to get current state for client_event broadcast.
  ClientEventData getFullClientEvent() => localPlayer.getFullClientEvent();

  // ---------------------------------------------------------------------------

  @override
  void dispose() {
    localPlayer.removeListener(_onSourceChanged);
    remotePlayer.removeListener(_onSourceChanged);
    super.dispose();
  }
}
